ArduCopter/system.cpp

```diff
@@ -304,8 +305,34 @@ void Copter::update_dynamic_notch()
         }
 #endif
     } else {
-        // set the harmonic notch filter frequency approximately scaled on motor rpm implied by throttle
-        ins.update_harmonic_notch_freq_hz(ref_freq * MAX(1.0f, sqrtf(motors->get_throttle_out() / ref)));
+        switch (ins.get_gyro_harmonic_notch_tracking_mode()) {
+#ifdef HAVE_AP_BLHELI_SUPPORT
+        case HarmonicNotch_UpdateBLHeli: { // BLHeli-based tracking
+            AP_BLHeli *blheli = AP_BLHeli::get_singleton();
+
+            if (blheli && blheli->get_num_motors() > 0) {
+                AP_BLHeli::telem_data td;
+                float motor_freq = 0.0f;
+                // average the rpm of each motor as reported by BLHeli and convert to Hz
+                for (uint8_t i = 0; i < blheli->get_num_motors(); i++) {
```

Comment:
it would be worth checking the age of the telem data, and perhaps a check for the RPM being in a reasonable range.
Also, should call get_num_motors() only once into a local var

Reply:
Yes and doing this inside of AP_BLHeli makes it all a lot cleaner - so thanks, fixed. Too low is covered by the MAX check in Copter::update_dynamic_notch(), too high is covered in the notch filter code which will prevent setting anything above the nyquist frequency.